daemons/pacemakerd/pcmkd_ipc.c

```diff
@@ -14,60 +14,102 @@
 
 #include "pacemakerd.h"
 
+/*!
+ * \internal
+ * \brief Accept a new client IPC connection
+ *
+ * \param[in,out] c    New connection
+ * \param[in]     uid  Client user id
+ * \param[in]     gid  Client group id
+ *
+ * \return 0 on success, -errno otherwise
+ */
 static int32_t
-pacemakerd_ipc_accept(qb_ipcs_connection_t * c, uid_t uid, gid_t gid)
+pacemakerd_ipc_accept(qb_ipcs_connection_t *c, uid_t uid, gid_t gid)
 {
-    crm_trace("Connection %p", c);
+    crm_trace("New client connection %p", c);
     if (pcmk__new_client(c, uid, gid) == NULL) {
         return -ENOMEM;
     }
     return 0;
 }
 
-/* Error code means? */
+/*!
+ * \internal
+ * \brief Destroy a client IPC connection
+ *
+ * \param[in] c  Connection to destroy
+ *
+ * \return 0 (i.e. do not re-run this callback)
+ */
 static int32_t
-pacemakerd_ipc_closed(qb_ipcs_connection_t * c)
+pacemakerd_ipc_closed(qb_ipcs_connection_t *c)
 {
     pcmk__client_t *client = pcmk__find_client(c);
 
     if (client == NULL) {
-        return 0;
-    }
-    crm_trace("Connection %p", c);
-    if (shutdown_complete_state_reported_to == client->pid) {
-        shutdown_complete_state_reported_client_closed = TRUE;
-        if (shutdown_trigger) {
-            mainloop_set_trigger(shutdown_trigger);
+        crm_trace("Ignoring request to clean up unknown connection %p", c);
+    } else {
+        crm_trace("Cleaning up closed client connection %p", c);
+
+        if (shutdown_complete_state_reported_to == client->pid) {
+            shutdown_complete_state_reported_client_closed = true;
+
+            if (shutdown_trigger) {
+                mainloop_set_trigger(shutdown_trigger);
+            }
         }
+
+        pcmk__free_client(client);
     }
-    pcmk__free_client(client);
+
     return 0;
 }
 
+/*!
+ * \internal
+ * \brief Destroy a client IPC connection
+ *
+ * \param[in,out] c  Connection to destroy
```

Comment:
Probably should be `[in]` -- see https://github.com/ClusterLabs/pacemaker/pull/3996/commits/ca584fb3a761b07ddfef0e7360bb446a21896992#r2559172284